### ISSUE TITLE
`std::variant` with alternative with `constexpr` destructor is not a literal type

### ISSUE THREAD

**Issue by JohelEGP** [llvm/llvm-project#56822]
This prevents useful by-value `constexpr` functions: https://godbolt.org/z/5ovrYzjdT.
```C++
#include <variant>
struct X { constexpr ~X() { } };
constexpr void f(std::variant<X>) { }
```
```console
<source>:3:33: error: constexpr function's 1st parameter type 'std::variant<X>' is not a literal type
constexpr void f(std::variant<X>) { }
                 ~~~~~~~~~~~~~~~^
/opt/compiler-explorer/clang-trunk-20220730/bin/../include/c++/v1/variant:1467:39: note: 'variant<X>' is not literal because it has data member '__impl' of non-literal type '__variant_detail::__impl<X>'
  __variant_detail::__impl<_Types...> __impl;
                                      ^
```

**Comment by JohelEGP** [comment]
And much more: https://godbolt.org/z/a9bcK8ob7.
```C++
#include <variant>
struct X { constexpr ~X() { } };
struct Y { std::variant<X> xyz; };
constexpr Y y{};
```
```
<source>:4:13: error: constexpr variable cannot have non-literal type 'const Y'
constexpr Y y{};
            ^
<source>:3:28: note: 'Y' is not literal because it has data member 'xyz' of non-literal type 'std::variant<X>'
struct Y { std::variant<X> xyz; };
                           ^
```

**Comment by frederick-vs-ja** [comment]
Currently the `variant` part of [P2231R1](https://wg21.link/p2231) is unimplemented in libc++. See also [this note](https://libcxx.llvm.org/Status/Cxx20.html#note-p2231).

@EugeneZelenko I think clang frontend is no problem here, and only libc++ needs update.

**Comment by JohelEGP** [comment]
I did notice. But since it already works for trivial types, and the standard doesn't mandate that an specialization of `std::variant` be a literal type, I filed this. I guess my issue really lies with https://wg21.link/LWG2833.

**Comment by frederick-vs-ja** [comment]
> I did notice. But since it already works for trivial types, and the standard doesn't mandate that an specialization of `std::variant` be a literal type, I filed this. I guess my issue really lies with https://wg21.link/LWG2833.

I think C++17 (or C++17 with DR [P0602R4](https://wg21.link/p0602r4)) has already mandated a `std::variant` specialization to be a literal type if all of its alternatives are trivially destructible (by `constexpr` constructors and requiring its destructor to be trivial). Some aspects of `constexpr` in standard library are currently underspecified, but IIUC they don't affect whether a certain type is literal.

**Comment by JohelEGP** [comment]
Add https://eel.is/c++draft/objects.within.classes#1 to the underspecification of `constexpr` in the C++ standard library, and I think it's underspecified whether `std::variant<std::vector<int>>` is required to be a literal type.

**Comment by frederick-vs-ja** [comment]
Should have been fixed by #83335.
